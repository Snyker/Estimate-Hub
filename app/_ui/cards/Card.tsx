import React, {PropsWithChildren} from "react";

export const Card: React.FC<PropsWithChildren<React.DetailedHTMLProps<React.HTMLAttributes<HTMLDivElement>, HTMLDivElement>>> = ({className='',children,...props}) => {
    return (
        <div className={`card max-w-none hover:scale-105 active:scale-95 hover:bg-gray-5 transition ease-in-out cursor-pointer ${className}`} {...props}>
            {children}
        </div>
    );
};
