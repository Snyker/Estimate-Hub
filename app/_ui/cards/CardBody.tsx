import React, {PropsWithChildren} from "react";

export const CardBody: React.FC<PropsWithChildren<React.DetailedHTMLProps<React.HTMLAttributes<HTMLDivElement>, HTMLDivElement>>> = ({className='',children,...props}) => {
    return (
        <div className={`card-body select-none ${className}`} {...props}>
            {children}
        </div>
    );
};
