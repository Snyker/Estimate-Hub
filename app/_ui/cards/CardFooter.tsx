import React, {PropsWithChildren} from "react";

export const CardFooter: React.FC<PropsWithChildren<React.DetailedHTMLProps<React.HTMLAttributes<HTMLDivElement>, HTMLDivElement>>> = ({className='flex-col items-start',children,...props}) => {
    return (
        <div className={`card-footer ${className}`} {...props}>
            {children}
        </div>
    );
};
