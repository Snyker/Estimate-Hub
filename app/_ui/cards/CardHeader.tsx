import React, {PropsWithChildren} from "react";

export const CardHeader: React.FC<PropsWithChildren<React.DetailedHTMLProps<React.HTMLAttributes<HTMLDivElement>, HTMLDivElement>>> = ({className='',children,...props}) => {
    return (
        <div className={`card-header justify-start gap-4 ${className}`} {...props}>
            {children}
        </div>
    );
};
