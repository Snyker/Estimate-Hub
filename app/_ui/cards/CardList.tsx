import React, {HTMLAttributes, PropsWithChildren} from "react";

export const CardList: React.FC<PropsWithChildren<HTMLAttributes<HTMLDivElement>>> = ({ className='', children, ...props}) => {
    return (
        <div className={`grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 2xl:grid-cols-4 gap-4 ${className}`} {...props}>
            {children}
        </div>
    );
};

