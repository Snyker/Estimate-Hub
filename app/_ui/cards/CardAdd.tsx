import {Card} from "@/app/_ui/cards/Card";
import {CardBody} from "@/app/_ui/cards/CardBody";
import {FaPlus} from "react-icons/fa6";
import {CardHeader} from "@/app/_ui/cards/CardHeader";
import React from "react";

export const CardAdd: React.FC<CardAddProps> = ({ title, onClick }) => {
    return (
        <Card onClick={onClick}>
            <CardBody className={'h-full justify-center items-center'}>
                <CardHeader className={'text-center'}>
                    <span>{title}</span>
                </CardHeader>
                <FaPlus size={64}/>
            </CardBody>
        </Card>
    );
};

export type CardAddProps = {
    title: string;
    onClick?: (() => void) | undefined;
}
