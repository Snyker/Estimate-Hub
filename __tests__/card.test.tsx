import React from 'react';
import renderer from 'react-test-renderer';
import { act } from 'react-dom/test-utils';
import {Card} from "@/app/_ui/cards/Card";
import {CardAdd, CardAddProps} from "@/app/_ui/cards/CardAdd";
import {CardHeader} from "@/app/_ui/cards/CardHeader";
import {CardBody} from "@/app/_ui/cards/CardBody";
import {CardFooter} from "@/app/_ui/cards/CardFooter";

describe('Card Components', () => {
    const defaultProps = {
        children: 'Card content',
    };

    describe('Card', () => {
        it('renders correctly with default props', () => {
            const tree = renderer.create(<Card {...defaultProps} />);
            expect(tree.toJSON()).toMatchSnapshot();
        });

        it('applies custom class', () => {
            const tree = renderer.create(<Card {...defaultProps} className="custom-class" />);
            expect(tree.toJSON()).toMatchSnapshot();
        });

        it('handles onClick event', () => {
            const onClickMock = jest.fn();
            const tree = renderer.create(<Card {...defaultProps} onClick={onClickMock} />);
            act(() => {
                tree.root.props.onClick();
            });
            expect(onClickMock).toHaveBeenCalled();
        });

        it('renders content based on children', () => {
            const tree = renderer.create(<Card {...defaultProps}><div>Custom content</div></Card>);
            expect(tree.toJSON()).toMatchSnapshot();
        });

        it('renders additional HTML attributes', () => {
            const tree = renderer.create(<Card {...defaultProps} data-testid="card" />);
            expect(tree.toJSON()).toMatchSnapshot();
        });

        it('renders correctly with multiple complex props', () => {
            const tree = renderer.create(
                <Card {...defaultProps} className="custom-class" onClick={() => {}} data-testid="card">
                    <div>Custom content</div>
                </Card>
            );
            expect(tree.toJSON()).toMatchSnapshot();
        });
    });

    describe('CardAdd', () => {
        const cardAddProps: CardAddProps = {
            title: 'Add Card',
        };

        it('renders correctly with default props', () => {
            const tree = renderer.create(<CardAdd {...cardAddProps} />);
            expect(tree.toJSON()).toMatchSnapshot();
        });

        it('renders additional HTML attributes', () => {
            const tree = renderer.create(<CardAdd {...cardAddProps} data-testid="card-add" />);
            expect(tree.toJSON()).toMatchSnapshot();
        });

    });

    describe('CardBody', () => {
        it('renders correctly with default props', () => {
            const tree = renderer.create(<CardBody {...defaultProps} />);
            expect(tree.toJSON()).toMatchSnapshot();
        });

        it('applies custom class', () => {
            const tree = renderer.create(<CardBody {...defaultProps} className="custom-class" />);
            expect(tree.toJSON()).toMatchSnapshot();
        });

        it('handles onClick event', () => {
            const onClickMock = jest.fn();
            const tree = renderer.create(<CardBody {...defaultProps} onClick={onClickMock} />);
            act(() => {
                tree.root.props.onClick();
            });
            expect(onClickMock).toHaveBeenCalled();
        });

        it('renders content based on children', () => {
            const tree = renderer.create(<CardBody {...defaultProps}><div>Custom content</div></CardBody>);
            expect(tree.toJSON()).toMatchSnapshot();
        });

        it('renders additional HTML attributes', () => {
            const tree = renderer.create(<CardBody {...defaultProps} data-testid="card-body" />);
            expect(tree.toJSON()).toMatchSnapshot();
        });

        it('renders correctly with multiple complex props', () => {
            const tree = renderer.create(
                <CardBody {...defaultProps} className="custom-class" onClick={() => {}} data-testid="card-body">
                    <div>Custom content</div>
                </CardBody>
            );
            expect(tree.toJSON()).toMatchSnapshot();
        });
    });

    describe('CardFooter', () => {
        it('renders correctly with default props', () => {
            const tree = renderer.create(<CardFooter {...defaultProps} />);
            expect(tree.toJSON()).toMatchSnapshot();
        });

        it('applies custom class', () => {
            const tree = renderer.create(<CardFooter {...defaultProps} className="custom-class" />);
            expect(tree.toJSON()).toMatchSnapshot();
        });

        it('renders content based on children', () => {
            const tree = renderer.create(<CardFooter {...defaultProps}><div>Custom content</div></CardFooter>);
            expect(tree.toJSON()).toMatchSnapshot();
        });

        it('renders additional HTML attributes', () => {
            const tree = renderer.create(<CardFooter {...defaultProps} data-testid="card-footer" />);
            expect(tree.toJSON()).toMatchSnapshot();
        });

        it('renders correctly with multiple complex props', () => {
            const tree = renderer.create(
                <CardFooter {...defaultProps} className="custom-class" data-testid="card-footer">
                    <div>Custom content</div>
                </CardFooter>
            );
            expect(tree.toJSON()).toMatchSnapshot();
        });
    });

    describe('CardHeader', () => {
        it('renders correctly with default props', () => {
            const tree = renderer.create(<CardHeader {...defaultProps} />);
            expect(tree.toJSON()).toMatchSnapshot();
        });

        it('applies custom class', () => {
            const tree = renderer.create(<CardHeader {...defaultProps} className="custom-class" />);
            expect(tree.toJSON()).toMatchSnapshot();
        });

        it('renders content based on children', () => {
            const tree = renderer.create(<CardHeader {...defaultProps}><div>Custom content</div></CardHeader>);
            expect(tree.toJSON()).toMatchSnapshot();
        });

        it('renders additional HTML attributes', () => {
            const tree = renderer.create(<CardHeader {...defaultProps} data-testid="card-header" />);
            expect(tree.toJSON()).toMatchSnapshot();
        });

        it('renders correctly with multiple complex props', () => {
            const tree = renderer.create(
                <CardHeader {...defaultProps} className="custom-class" data-testid="card-header">
                    <div>Custom content</div>
                </CardHeader>
            );
            expect(tree.toJSON()).toMatchSnapshot();
        });
    });
});
