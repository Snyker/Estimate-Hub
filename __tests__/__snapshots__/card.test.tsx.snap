// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Components Card applies custom class 1`] = `
<div
  className="card hover:scale-105 active:scale-95 hover:bg-gray-5 transition ease-in-out cursor-pointer md:max-h-[14rem] custom-class"
>
  Card content
</div>
`;

exports[`Card Components Card renders additional HTML attributes 1`] = `
<div
  className="card hover:scale-105 active:scale-95 hover:bg-gray-5 transition ease-in-out cursor-pointer md:max-h-[14rem] "
  data-testid="card"
>
  Card content
</div>
`;

exports[`Card Components Card renders content based on children 1`] = `
<div
  className="card hover:scale-105 active:scale-95 hover:bg-gray-5 transition ease-in-out cursor-pointer md:max-h-[14rem] "
>
  <div>
    Custom content
  </div>
</div>
`;

exports[`Card Components Card renders correctly with default props 1`] = `
<div
  className="card hover:scale-105 active:scale-95 hover:bg-gray-5 transition ease-in-out cursor-pointer md:max-h-[14rem] "
>
  Card content
</div>
`;

exports[`Card Components Card renders correctly with multiple complex props 1`] = `
<div
  className="card hover:scale-105 active:scale-95 hover:bg-gray-5 transition ease-in-out cursor-pointer md:max-h-[14rem] custom-class"
  data-testid="card"
  onClick={[Function]}
>
  <div>
    Custom content
  </div>
</div>
`;

exports[`Card Components CardAdd renders additional HTML attributes 1`] = `
<div
  className="card hover:scale-105 active:scale-95 hover:bg-gray-5 transition ease-in-out cursor-pointer md:max-h-[14rem] "
>
  <div
    className="card-body h-full justify-center items-center"
  >
    <div
      className="card-header "
    >
      <span>
        Add Card
      </span>
    </div>
    <svg
      fill="currentColor"
      height={64}
      stroke="currentColor"
      strokeWidth="0"
      style={
        {
          "color": undefined,
        }
      }
      viewBox="0 0 448 512"
      width={64}
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M256 80c0-17.7-14.3-32-32-32s-32 14.3-32 32V224H48c-17.7 0-32 14.3-32 32s14.3 32 32 32H192V432c0 17.7 14.3 32 32 32s32-14.3 32-32V288H400c17.7 0 32-14.3 32-32s-14.3-32-32-32H256V80z"
      />
    </svg>
  </div>
</div>
`;

exports[`Card Components CardAdd renders correctly with default props 1`] = `
<div
  className="card hover:scale-105 active:scale-95 hover:bg-gray-5 transition ease-in-out cursor-pointer md:max-h-[14rem] "
>
  <div
    className="card-body h-full justify-center items-center"
  >
    <div
      className="card-header "
    >
      <span>
        Add Card
      </span>
    </div>
    <svg
      fill="currentColor"
      height={64}
      stroke="currentColor"
      strokeWidth="0"
      style={
        {
          "color": undefined,
        }
      }
      viewBox="0 0 448 512"
      width={64}
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M256 80c0-17.7-14.3-32-32-32s-32 14.3-32 32V224H48c-17.7 0-32 14.3-32 32s14.3 32 32 32H192V432c0 17.7 14.3 32 32 32s32-14.3 32-32V288H400c17.7 0 32-14.3 32-32s-14.3-32-32-32H256V80z"
      />
    </svg>
  </div>
</div>
`;

exports[`Card Components CardBody applies custom class 1`] = `
<div
  className="card-body custom-class"
>
  Card content
</div>
`;

exports[`Card Components CardBody renders additional HTML attributes 1`] = `
<div
  className="card-body "
  data-testid="card-body"
>
  Card content
</div>
`;

exports[`Card Components CardBody renders content based on children 1`] = `
<div
  className="card-body "
>
  <div>
    Custom content
  </div>
</div>
`;

exports[`Card Components CardBody renders correctly with default props 1`] = `
<div
  className="card-body "
>
  Card content
</div>
`;

exports[`Card Components CardBody renders correctly with multiple complex props 1`] = `
<div
  className="card-body custom-class"
  data-testid="card-body"
  onClick={[Function]}
>
  <div>
    Custom content
  </div>
</div>
`;

exports[`Card Components CardFooter applies custom class 1`] = `
<div
  className="card-footer custom-class"
>
  Card content
</div>
`;

exports[`Card Components CardFooter renders additional HTML attributes 1`] = `
<div
  className="card-footer flex-col items-start"
  data-testid="card-footer"
>
  Card content
</div>
`;

exports[`Card Components CardFooter renders content based on children 1`] = `
<div
  className="card-footer flex-col items-start"
>
  <div>
    Custom content
  </div>
</div>
`;

exports[`Card Components CardFooter renders correctly with default props 1`] = `
<div
  className="card-footer flex-col items-start"
>
  Card content
</div>
`;

exports[`Card Components CardFooter renders correctly with multiple complex props 1`] = `
<div
  className="card-footer custom-class"
  data-testid="card-footer"
>
  <div>
    Custom content
  </div>
</div>
`;

exports[`Card Components CardHeader applies custom class 1`] = `
<div
  className="card-header custom-class"
>
  Card content
</div>
`;

exports[`Card Components CardHeader renders additional HTML attributes 1`] = `
<div
  className="card-header "
  data-testid="card-header"
>
  Card content
</div>
`;

exports[`Card Components CardHeader renders content based on children 1`] = `
<div
  className="card-header "
>
  <div>
    Custom content
  </div>
</div>
`;

exports[`Card Components CardHeader renders correctly with default props 1`] = `
<div
  className="card-header "
>
  Card content
</div>
`;

exports[`Card Components CardHeader renders correctly with multiple complex props 1`] = `
<div
  className="card-header custom-class"
  data-testid="card-header"
>
  <div>
    Custom content
  </div>
</div>
`;
